bootstrap/util/configowner.go

```diff
@@ -47,6 +47,30 @@ func (co ConfigOwner) IsInfrastructureReady() bool {
 	return infrastructureReady
 }
 
+// NodeHasJoined checks if the config owner has a nodeRef for each machine.
+func (co ConfigOwner) NodeHasJoined() bool {
+	if co.IsMachinePool() {
+		numExpectedNodes, found, err := unstructured.NestedInt64(co.Object, "spec", "replicas")
+		if err != nil {
+			return false
+		}
+		// replicas default to 1 so this is what we should use if nothing is specified
+		if !found {
+			numExpectedNodes = 1
+		}
+		nodeRefs, _, err := unstructured.NestedSlice(co.Object, "status", "nodeRefs")
+		if err != nil {
+			return false
+		}
+		return len(nodeRefs) == int(numExpectedNodes)
+	}
```

Comment:
This change makes sense for MachinePools as well, ideally we don't want to rotate the token mid-node join (for example if the machine pool instance has already been created but the node hasn't joined yet), we want to extend the current token so it can be consumed just like Machines. Then, once the nodes have joined, we rotate the token periodically to be ready for the next replica increase.

My question is how does `NodeHasJoined` apply to MachinePools? MachinePools don't map to a single node, they have a node ref list.

Reply:
> My question is how does NodeHasJoined apply to MachinePools? MachinePools don't map to a single node, they have a node ref list.

@CecileRobertMichon true. Maybe I should rename it, perhaps `HasNodeRefs`, or a separate function for MachinePools? Now it is checking that a MachinePool has as many nodeRefs as there are replicas (and that a Machine has a nodeRef). I'm not sure if this is the best thing to check though and would be grateful for insights.

Reply:
@CecileRobertMichon if we expect token to stick around for MachinePools - thus halting rotation for a while - then I assume we must change also something in the code I have linked above because as far as I understand it the current implementation rotates token when TTL approaches (no matter of the node joining)

Reply:
> Maybe I should rename it, perhaps HasNodeRefs, or a separate function for MachinePools

+1 to either renaming or separating the func in two. Making sure there are as many nodeRefs as spec Replicas makes sense for MachinePools to check that all desired nodes have joined.

>  I assume we must change also something in the code I have linked above because as far as I understand it the current implementation rotates token when TTL approaches (no matter of the node joining)

The code is correct: it only extends the token when TTL approaches if and only if the nodes have joined (or prior to this change if infra is ready), as the first `if` statement returns out of the func if we need to extend the token to wait for nodes/infra to join. 

https://github.com/kubernetes-sigs/cluster-api/pull/6395/files#diff-1bdaceb68bfe880940139bae93a34f96a05461ee4dd0fa99f8374928c00242fbR252

Reply:
> +1 to either renaming or separating the func in two. Making sure there are as many nodeRefs as spec Replicas makes sense for MachinePools to check that all desired nodes have joined.

Renamed to `HasNodeRefs` and elaborated a bit in the docstring. :slightly_smiling_face:

Reply:
Ahh, 
```
return r.refreshBootstrapToken
```
Got it, sorry if I asked twice but somehow I missed this detail
thanks, @CecileRobertMichon for clarifying